source/extensions/filters/http/common/aws/signer_impl.cc

```diff
@@ -0,0 +1,256 @@
+  if (path.empty()) {
+    out << "/";
+  } else {
+    out << path;
+  }
+  out << "\n";
+  // Query string
+  const auto query = StringUtil::cropLeft(path_value.getStringView(), "?");
+  if (query != path) {
+    out << query;
+  }
+  out << "\n";
+  // Headers
+  for (const auto& header : canonical_headers) {
+    out << header.first << ":" << header.second << "\n";
+  }
+  out << "\n" << signing_headers << "\n";
+  // Content Hash
+  out << content_hash;
+  return out.str();
+}
+
+std::string SignerImpl::createSigningHeaders(
+    const std::map<std::string, std::string>& canonical_headers) const {
+  std::vector<std::string> keys;
+  keys.reserve(canonical_headers.size());
+  for (const auto& header : canonical_headers) {
+    keys.emplace_back(header.first);
+  }
+  return StringUtil::join(keys, ";");
+}
+
+std::string SignerImpl::createCredentialScope(const std::string& short_date,
+                                              const std::string& region) const {
+  std::stringstream out;
+  out << short_date << "/" << region << "/" << service_name_ << "/" << AWS4_REQUEST;
+  return out.str();
+}
+
+std::string SignerImpl::createStringToSign(const std::string& canonical_request,
+                                           const std::string& long_date,
+                                           const std::string& credential_scope) const {
+  std::stringstream out;
+  out << AWS4_HMAC_SHA256 << "\n";
+  out << long_date << "\n";
+  out << credential_scope << "\n";
+  out << Hex::encode(hash(Buffer::OwnedImpl(canonical_request)));
+  return out.str();
+}
+
+std::string SignerImpl::createSignature(const std::string& secret_access_key,
+                                        const std::string& short_date, const std::string& region,
+                                        const std::string& string_to_sign) const {
+  const auto k_secret = AWS4 + secret_access_key;
+  const auto k_date = hmac(std::vector<uint8_t>(k_secret.begin(), k_secret.end()), short_date);
+  const auto k_region = hmac(k_date, region);
+  const auto k_service = hmac(k_region, service_name_);
+  const auto k_signing = hmac(k_service, AWS4_REQUEST);
+  return Hex::encode(hmac(k_signing, string_to_sign));
+}
+
+std::string SignerImpl::createAuthorizationHeader(const std::string& access_key_id,
+                                                  const std::string& credential_scope,
+                                                  const std::string& signing_headers,
+                                                  const std::string& signature) const {
+  std::stringstream out;
+  out << AWS4_HMAC_SHA256 << " ";
+  out << CREDENTIAL << "=" << access_key_id << "/" << credential_scope << ", ";
+  out << SIGNED_HEADERS << "=" << signing_headers << ", ";
+  out << SIGNATURE << "=" << signature;
+  return out.str();
+}
+
+std::map<std::string, std::string>
+SignerImpl::canonicalizeHeaders(const Http::HeaderMap& headers) const {
```

Comment:
The logic in here is quite arcane; do you think we could split this out of a private method into a standalone utility and write some extensive tests for this? It seems a possible source of compliance bugs.